File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/support/ShardInfo.java
package fr.myprysm.vertx.elasticsearch.action.support;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class ShardInfo {
    private int total;
    private int successful;
    private int failed;
    private List<ShardInfoFailure> failures;


    public ShardInfo(ShardInfo other) {
        total = other.total;
        successful = other.successful;
        failures = other.failures;
        failed = other.failed;
    }

    public ShardInfo(JsonObject json) {
        ShardInfoConverter.fromJson(json, this);
    }

    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        ShardInfoConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/support/ShardInfo.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.support

import fr.myprysm.vertx.elasticsearch.action.support.ShardInfo
import fr.myprysm.vertx.elasticsearch.action.support.ShardInfoFailure

fun ShardInfo(
  failed: Int? = null,
  failures: Iterable<fr.myprysm.vertx.elasticsearch.action.support.ShardInfoFailure>? = null,
  successful: Int? = null,
  total: Int? = null): ShardInfo = fr.myprysm.vertx.elasticsearch.action.support.ShardInfo().apply {

  if (failed != null) {
    this.setFailed(failed)
  }
  if (failures != null) {
    this.setFailures(failures.toList())
  }
  if (successful != null) {
    this.setSuccessful(successful)
  }
  if (total != null) {
    this.setTotal(total)
  }
}

